lib/core/test/unit/Cardano/Wallet/Primitive/CoinSelection/RandomSpec.hs

```diff
@@ -151,15 +160,106 @@ spec = do
                 , validateSelection = noValidation
                 , utxoInputs = [oneAda, oneAda, oneAda, oneAda]
                 , txOutputs = 2*oneAda :| [oneAda `div` 2]
+                , totalWithdrawal = 0
                 })
 
+        coinSelectionUnitTest random "withdrawal simple"
+            (Right $ CoinSelectionResult
+                { rsInputs = [1]
+                , rsChange = []
+                , rsOutputs = [2]
+                })
+            (CoinSelectionFixture
+                { maxNumOfInputs = 100
+                , validateSelection = noValidation
+                , utxoInputs = [1]
+                , txOutputs = 2 :| []
+                , totalWithdrawal = 1
+                })
+
+        coinSelectionUnitTest random "withdrawal multi-output"
+            (Right $ CoinSelectionResult
+                { rsInputs = [1,1]
+                , rsChange = []
+                , rsOutputs = [2,2]
+                })
+            (CoinSelectionFixture
+                { maxNumOfInputs = 100
+                , validateSelection = noValidation
+                , utxoInputs = [1,1]
+                , txOutputs = 2 :| [2]
+                , totalWithdrawal = 2
+                })
+
+        coinSelectionUnitTest random "withdrawal not even"
+            -- 10 Ada available as withdrawal
+            --
+            -- - 7 Ada goes to the first output (10/14 * 10 ~= 7)
+            -- - 2 Ada goes to the other output ( 4/14 * 10 ~= 2)
+            (Right $ CoinSelectionResult
+                { rsInputs = [5,5]
+                , rsChange = [2,3]
+                , rsOutputs = [10, 4]
+                })
+            (CoinSelectionFixture
+                { maxNumOfInputs = 100
+                , validateSelection = noValidation
+                , utxoInputs = [5,5]
+                , txOutputs = 10 :| [4]
+                , totalWithdrawal = 10
+                })
+
+        coinSelectionUnitTest random "withdrawal cover next output"
+            -- 20 Ada available as withdrawal
+            --
+            -- - 10 Ada goes to the first output
+            -- - 10 Ada goes to the other output
+            --
+            -- The first output has to select an available input first, which
+            -- leaves no input for the second, but it can be covered with the
+            -- withdrawal.
+            (Right $ CoinSelectionResult
+                { rsInputs = [1]
+                , rsChange = [1]
+                , rsOutputs = [10, 10]
+                })
+            (CoinSelectionFixture
+                { maxNumOfInputs = 100
+                , validateSelection = noValidation
+                , utxoInputs = [1]
+                , txOutputs = 10 :| [10]
```

Comment:
in old days it would trigger not enough fragmentation if the number of inputs does not match number of outputs. now I see withdrawals help in this sense :+1:  wonder if there are some integration tests checking this edge conditions that needs to be updated

Reply:
what about this example : 
```
            (CoinSelectionFixture
                { maxNumOfInputs = 100
                , validateSelection = noValidation
                , utxoInputs = [1]
                , txOutputs = 1 :| [1,1,1,1,1]
               , totalWithdrawal = 5
                })
``` ?

Reply:
> what about this example

Following the _current_ logic, each output would receive `0` withdrawal bonus (1/6 * 5, rounded down) and therefore, this would fail on the second output because there wouldn't be enough inputs to select. Although indeed, if instead of using the withdrawal proportionally, we were using them fully to cover "just" the right amount, it'd give better result.